feat: make cypress.config.js available

elevatebart opened the pull request:
closes 
https://cypress-io.atlassian.net/browse/UNIFY-18
https://cypress-io.atlassian.net/browse/UNIFY-208
https://cypress-io.atlassian.net/browse/UNIFY-209
https://cypress-io.atlassian.net/browse/UNIFY-247
https://cypress-io.atlassian.net/browse/UNIFY-248
https://cypress-io.atlassian.net/browse/UNIFY-249
https://cypress-io.atlassian.net/browse/UNIFY-250
https://cypress-io.atlassian.net/browse/UNIFY-17
https://cypress-io.atlassian.net/browse/UNIFY-264
https://cypress-io.atlassian.net/browse/UNIFY-212

closes #5218

## Summary

In preparation for the unified runner project, we need to be able to detect which testing types have been setup.
For this detection, we plan on using `e2e` and `component` function. So we need those functions to be there before unification.


## ~~cypress.json~~ cypress.config.ts

To configure cypress, we can now use `cypress.config.ts` or `cypress.config.js` instead of `cypress.json`. 
Using a `ts` or `js` file to store config needs cypress to execute/evaluate the file in a node context. 
Since this is user code, this could throw a parsing/runtime error and crash the electron runner. 
To avoid crashing, we spawn another process to parse the file. It allows cypress to elegantly handle each error. 

One downside of using a separate process is this: 
Functions in this config file cannot be transmitted to the main process. 
We don't get the new e2e and component functions in the config object passed around everywhere.
It matters very little since they are supposed to be run in another child process anyway (see run_plugin.js) but it is worth noting.

### Default config file

Three default config files are now possible: `cypress.json`, `cypress.config.js` and `cypress.config.ts`. If more than one of those files is present at the root of the project, the system errors because of abiguity.

### Creating missing file

Before, when the targeted config file did not exist cypress would create it and populate it with the arguments passed to the CLI.
It still does that if a json config file is specified in the CLI. It creates a config file by default if a `js` or `ts` file is specified and not found. By default, it's going to create a `cypress.config.js` file.

## Plugins definition functions

Before, all the servers and preprocessors had to be defined in the `pluginsFile`. This js/ts file whose path was configured in `cypress.json` was executed when loading a project on the cypress server. It allowed users to modify the cypress config depending on execution context and setup preprocessors, tasks and event handlers. It exported a function like below.

```js
const { startDevServer } = require('@cypress/webpack-dev-server')

module.exports = (on, config) => {
  on('dev-server:start', (options) => {
    return startDevServer({ options, webpackConfig })
  })

  require('@cypress/code-coverage/task')(on, config)
  return config
}
```

Though we can still do that, now we can set this up directly in the `cypress.config.ts`.
Note how we setup the dev-server for component testing as a separate function `setupDevServer`.

```ts
export default {
    // ...
    component: {
        setupDevServer (options) {
            return startDevServer({ options, webpackConfig })
        },
        setupNodeEvents (on, config) {
             require('@cypress/code-coverage/task')(on, config)
             return config
        }
    }
}
```

If a plugins function and a pluginsFile are implemented, the system will throw an error. We want to avoid halfway migration.

### Technical implementation

When it gets time to execute the `pluginsFile`, if `config.component` is a function, 
we replace the configured `pluginsFile` with `configFilePath` and tell the plugins resolver what function it is supposed to execute.
Else, I let the pluginsFile run as before.

For now, if neither the `pluginsFile` nor the functions in `cypress.config.ts` exist, cypress creates sample plugins file. This avoids breaking tests but encourages the creation of a deprecated configuration.

This should be avoided.

## defineConfig()

Vite, Vue CLI 4, windicss and a few other libraries have implemented a pattern for configuring their APIs that is very practical: a `defineConfig()` function. 
This enables autocompletion and inline documentation in the config file while minimizing boilerplate.
Here is how to use it.

```ts
import { defineConfig } from "cypress"

export default defineConfig({
    baseUrl: "./substack/"
})
```

<img width="548" alt="defineConfig" src="https://user-images.githubusercontent.com/5592465/125000464-14ee5c80-e016-11eb-8765-b017a9f459ca.png">


## desktop-gui

As a consequence of these changes, the desktop-gui was displaying some wrong info in the settings tab.

Here are my fixes:

<img width="1001" alt="Screen Shot 2021-07-08 at 5 30 12 PM" src="https://user-images.githubusercontent.com/5592465/125002232-13269800-e01a-11eb-877d-d3db7f479986.png">

## Deprecation warnings

To give a heads up about this necessary change. We add a warning both in the component testing runner and in the desktop GUI.

### To see the desktop warning:

- `yarn start`
- open "packages/example" project
- see the screenshot below

<img width="1269" alt="Screen Shot 2021-08-09 at 10 20 10 AM" src="https://user-images.githubusercontent.com/5592465/128730958-da2b0a28-be9d-41a3-8c50-94c4c760a8a0.png">

### To see the warning in the component testing runner

- `cd npm/react/examples/react-scripts`
- `yarn cy:open`
- see this screnshot

<img width="1304" alt="Screen Shot 2021-08-09 at 10 22 39 AM" src="https://user-images.githubusercontent.com/5592465/128731331-cdeff85d-17d7-49dc-a48f-9abeeb1cc51d.png">

### Technical detail

We had to wire up the component testing runner to receive warnings in the first place. 

Look in `server-ct/src/index.ts` to see how it was done.

## Remaining work before merge

These tasks remain once this change is approved. 
They have to be accomplished before we deliver this to users to give them a solution when they see the deprecation.

- [ ] Review text of errors and warnings
- [ ] Create and merge codemods in this branch
- [ ] Have a clear documentation ready to publish

Bkucera commented:
This might make migration to the new cypress.js (with plugins support) more difficult, causing users to do two config migrations back-to-back. Especially if we end up going with cypress.config.js. thoughts @cowboy @mccataldo ?

elevatebart commented:
@Bkucera I agree 💯 

- I will change `cypress.js` into `cypress.config.js`
- There is no plan to merge this in cypress 8 (yet)
- This change is non-breaking and does not involve the deprecation of anything

I am sorry for the initial confusion

chrisbreiding commented:
I share the concern that Ben mentioned. It might be more confusing to introduce one way of writing cypress.config.js and then another way than just to introduce the final way.

Also, a reason for utilizing the plugins file architecture for cypress.config.js is that it's contained in a child process and has more robust error handling. If there's an asynchronous error in this implementation of cypress.config.js, Cypress will crash. With the plugins file architecture, it would display an appropriately tailored error message.

> There is no plan to merge this in cypress 8 (yet)

What is the plan for releasing this?

cowboy commented:
> This might make migration to the new cypress.js (with plugins support) more difficult, causing users to do two config migrations back-to-back. Especially if we end up going with cypress.config.js. thoughts @cowboy @mccataldo ?

I think it would be wise to coordinate any proposed changes to `cypress.json` -> `cypress.js` so that we can minimize the number of changes we need to communicate to our users. It may be wise to hold off on merging this for the moment until we can understand if there will be other changes needed. And we definitely need to make sure that we agree on the name!

jennifer-shehane commented:
I'm also not understanding the purpose behind this change if it is not meant to be user facing before we make the changes to include plugin functionality to cypress.js. I read all of the Jira stories and the entire Notion doc and didn't see this as a step where we were headed. I may have missed it though.

lmiller1990 commented on npm/vue/cypress.config.js:
```diff
@@ -0,0 +1,32 @@
+const { startDevServer } = require('@cypress/webpack-dev-server')
```
Cool, in future we should try to use `git mv` to preserve the history. It will also make the PR smaller to review.

elevatebart replied:
In theory yes, but have you seen it actually work?

lmiller1990 replied:
Usually, but I had an experience recently where it did not - seems if you change too many LOC in the file, it doesn't work. Not really a big deal, but good to keep in mind.

lmiller1990 commented on npm/vue/cypress.config.js:
```diff
@@ -0,0 +1,32 @@
+  projectId: '134ej7',
+  testFiles: '**/*spec.js',
+  experimentalFetchPolyfill: true,
+  component: (on, config) => {
```
If we want to customize some values for component, eg viewport, what is the recommendation?

## Before

```json
{
  "component": {
    "viewportHeight": 400
  }
}
```

## After

```js
module.exports = {
  component: (on, config) => {
    // ...
    return {
      ...config,
      viewportHeight: 400
    }
  }
}
```

Is this correct? For anyone with overrides, the migration won't be as simple as rename + wrap in `module.exports`. We might want to include this in the PR description and (eventual) docs.

elevatebart replied:
I think this is how it was intended indeed.

This or a simple affectation (no need for the spread)

```js
module.exports = {
  component: (on, config) => {
    // ...
	config.viewportHeight = 400
    return config
  }
}
```

lmiller1990 commented on packages/server/test/cache_helper.js:
```diff
@@ -0,0 +1,19 @@
+  'cypress.config.js',
+]
+
+/**
+   * Since we load the cypress.json, cypress.e2e.config.js or cypress.component.config.js
+   * via `require`, we need to clear the `require.cache` before/after some tests
+   * to ensure we are not using a cached configuration file.
+   */
```
```suggestion
/**
 * Since we load the cypress.json, cypress.e2e.config.js or cypress.component.config.js
 * via `require`, we need to clear the `require.cache` before/after some tests
 * to ensure we are not using a cached configuration file.
 */
```

lmiller1990 commented on packages/server/test/unit/project_spec.js:
```diff
@@ -54,7 +55,7 @@ describe('lib/project-e2e', () => {
 
   afterEach(function () {
-    Fixtures.remove()
+    // Fixtures.remove()
```
```suggestion

```

lmiller1990 replied:
Why don't need this anymore?

lmiller1990 reviewed (commented):
Left some questions, but looks pretty solid!

jennifer-shehane commented on packages/driver/cypress/integration/commands/request_spec.js:
```diff
@@ -830,7 +830,7 @@ describe('src/cy/commands/request', () => {
           expect(lastLog.get('error')).to.eq(err)
           expect(lastLog.get('state')).to.eq('failed')
-          expect(err.message).to.eq('`cy.request()` must be provided a fully qualified `url` - one that begins with `http`. By default `cy.request()` will use either the current window\'s origin or the `baseUrl` in `cypress.json`. Neither of those values were present.')
+          expect(err.message).to.eq('`cy.request()` must be provided a fully qualified `url` - one that begins with `http`. By default `cy.request()` will use either the current window\'s origin or the `baseUrl` in `cypress.config.js`. Neither of those values were present.')
```
This error is slightly inaccurate since the baseUrl could have been passed via the --config outside of the configuration file. But, I know this wasn't introduced in this PR. 😝

elevatebart replied:
I fixed the error message to say option instead of the file, since it can com from either configfile or the CLI options but is most likely something wrong within the test.

jennifer-shehane reviewed (changes requested):
Nice work. Just mentioning a few edge cases I found. I didn't really test too extensively. I'm not sure if all of these issues are something that need addressing now or if you want to make separate tickets for future work. 

### Crash

I had Cypress crash at one point when I was editing the `cypress.config.js` file and now I can't reproduce it. 😞 There was no error in the stdout when it happened. 

### Export empty obj

If you export an empty object for the `cypress.config.js` file with a `cypress.json` file present - it sees this as if the `cypress.config.js` has been deleted completely and a misleading error displays. This should be supported and is likely the large majority of cases of how I see people would start the file when migrating.

```js
module.exports = {}
```

<img width="764" alt="Screen Shot 2021-06-29 at 11 32 47 AM" src="https://user-images.githubusercontent.com/1271364/123835221-e8c43480-d8cd-11eb-91b8-b865fe4545d5.png">

### nonexistent configFile

If I pass a nonexistent `--config-file` via `cypress open`, there's a pretty confusing error that displays in the GUI. The previous behavior was that it created a new file at the location. I actually don't remember that being the behavior or if that's what we want to continue doing. 

<img width="801" alt="Screen Shot 2021-06-29 at 11 46 37 AM" src="https://user-images.githubusercontent.com/1271364/123836997-f7135000-d8cf-11eb-8d46-ed63f1815771.png">

Additionally when I click to 'Try again' on this screen (still with the nonexistent file), Cypress gets stuck in a loading spinner and clicking into the other tabs just blank out - have to restart Cypress. 

<img width="799" alt="Screen Shot 2021-06-29 at 11 51 34 AM" src="https://user-images.githubusercontent.com/1271364/123837372-61c48b80-d8d0-11eb-9239-34f7fa48e5ab.png">

### Settings

The Settings tab is pretty wonky.   

#### e2e shows as null even though I have a function set

<img width="588" alt="Screen Shot 2021-06-29 at 11 21 50 AM" src="https://user-images.githubusercontent.com/1271364/123833917-79017a00-d8cc-11eb-9515-c723da6b056a.png">

<img width="632" alt="Screen Shot 2021-06-29 at 11 22 13 AM" src="https://user-images.githubusercontent.com/1271364/123833964-861e6900-d8cc-11eb-8fe5-420d4c36a84b.png">

#### These extra values are showing

These values don't show in 7.6.0.

<img width="632" alt="Screen Shot 2021-06-29 at 11 22 13 AM" src="https://user-images.githubusercontent.com/1271364/123834054-9cc4c000-d8cc-11eb-9f2e-740c6c463691.png">

#### configFile should show as relative path

<img width="771" alt="Screen Shot 2021-06-29 at 11 22 44 AM" src="https://user-images.githubusercontent.com/1271364/123834325-eb725a00-d8cc-11eb-9f8d-32eb56946456.png">

<img width="749" alt="Screen Shot 2021-06-29 at 11 27 13 AM" src="https://user-images.githubusercontent.com/1271364/123834367-f88f4900-d8cc-11eb-9f39-52e7b7f9621c.png">

#### The instructions for where to put projectId need updating

<img width="749" alt="Screen Shot 2021-06-29 at 11 27 13 AM" src="https://user-images.githubusercontent.com/1271364/123834689-515ee180-d8cd-11eb-8c79-a93ea842edd0.png">

#### Also Node.js instructions need updating

<img width="734" alt="Screen Shot 2021-06-29 at 11 30 40 AM" src="https://user-images.githubusercontent.com/1271364/123834861-7e12f900-d8cd-11eb-8994-7260af308a34.png">

### projectId

I'm curious on what would happen if someone exportts a different projectId for e2e versus component - will need to be considered and tested as I imagine someone would try to set up 2 different projects in the Dashboard.

jennifer-shehane commented:
Ah! I also saw this error in CI when building locally. I reran the build and it passed the next time, so doesn't seem isolated to CI. 

<img width="1005" alt="Screen Shot 2021-06-29 at 11 57 18 AM" src="https://user-images.githubusercontent.com/1271364/123838118-3c844d00-d8d1-11eb-9994-ececc213d699.png">

elevatebart commented:
@jennifer-shehane this one is on the main branch too I believe. It is linked to the configuration of typescript. I fixed it in my last commit.

lmiller1990 commented:
Great review above - can you tag me again when I should re-review this? Looks like you are still addressing some feedback and CI is red.

elevatebart commented:
@jennifer-shehane thank you for the great review.

I fixed the error when a config file was missing or misnamed.
I also fixed the error when a change in config triggered a crash by closing and re-opening the current project.

I am very hesitant to fix the `desktop-gui` settings tab, since it will need a lot of re-design
- What should I do with functions? (they cannot be transferred through ipc)
- Is cypress.env.json still relevant?
- Should we update the inline documentation twice? Once now, and once with the unified GUI?

cowboy commented on packages/server/lib/util/settings.js:
```diff
@@ -97,8 +135,37 @@ module.exports = {
+      return options.configFile
+    }
+
+    if (ls.includes('cypress.config.ts')) {
+      return 'cypress.config.ts'
+    }
+
+    if (ls.includes('cypress.config.js')) {
+      return 'cypress.config.js'
+    }
+
+    if (ls.includes('cypress.json')) {
+      return 'cypress.json'
+    }
```
It would be nice if there was a shared const like this that our code could use throughout

```js
const configFiles = ['cypress.config.ts', 'cypress.config.js', 'cypress.json']
```

So this could look like

```js
for (const configFile of configFiles) {
  if (ls.includes(configFile)) {
    return configFile
  }
}

```

elevatebart replied:
Done in 878035d1d5cbbded3ebbed1c88cffe4e49f2b81e

cowboy replied:
Did you intend to change the logic so that the `.js` file is now preferred over the `.ts` file?

elevatebart replied:
Actually, I even intend to error if I find `js` and `ts`.
We ultimately want to write in the config files. 
If we have a `js` and a `ts` which one do we read? which one do we write? Is it worth it to decide? I probably need to add this to this PR.

cowboy replied:
If we need to know the status of multiple config files, instead of the previously-mentioned `reduce` + `pathExistsAsync` code, we should do something more like this:

```js
Promise.all(CYPRESS_CONFIG_FILES.map(filename => {
  const cypressConfigPath = cwd(filename)
  return fs.pathExistsAsync(cypressConfigPath).then(found => ({
    filename,
    cypressConfigPath,
    exists
  }))
})).then(result => {
  // look at all entries in `result` to figure out how to handle cases where there
  // are both js + ts, etc

  // otherwise find the first matching one and use it
})
```

cowboy reviewed (changes requested):
I've made a few observations and comments. I hope they're helpful!

cowboy commented on cli/lib/cypress.js:
```diff
@@ -68,6 +68,10 @@ const cypressModuleApi = {
     },
   },
+
+  defineConfig (config) {
```
This function _definitely_ needs a comment explaining what's going on!

elevatebart replied:
true, I added a comment to guide devs to understand the meaninng of it.

elevatebart replied:
Done in 38e76acf70d4a9e3432c5f79a9cb02e41eae8d19

cowboy replied:
I don't think that's the right commit!

elevatebart commented on packages/server/lib/plugins/util.js:
```diff
@@ -4,11 +4,62 @@ const debug = require('debug')('cypress:server:plugins')
+      return arg.map((argElement) => serializeArgument(argElement))
+    }
+
+    return Object.keys(arg).reduce(function (acc, key) {
+      acc[key] = serializeArgument(arg[key])
+
+      return acc
+    }, {})
```
I agree that this would be less verbose, but it has 2 issue in my opinion:
- It is a lot more obscure than what is there right now
- It creates a new object at every iteration

I will try the Object.entries approach. I think it's a good one.

elevatebart replied:
```js
const serializedObject = {}

    for (const [key, val] of Object.entries(arg)) {
      serializedObject[key] = serializeArgument(val)
    }

    return serializedObject
```

elevatebart replied:
Done in 00c799d5f03467ce535043e459acd72c6d89ebab

elevatebart commented on packages/server/lib/util/settings.js:
```diff
@@ -3,7 +3,31 @@ const Promise = require('bluebird')
+const debug = require('debug')('cypress:server:settings')
+
+function jsCode (obj) {
+  const objJSON = obj && obj !== {} ? JSON.stringify(obj, null, 2) : `{
```
In e2e tests we (ab)use the _write function to setup a config file

cowboy replied:
But testing `obj && obj !== {}` is the same as saying `obj && true` because `obj` can never `===` an object defined inline `{}`, right? It seems like the `obj !== {}` bit is either unnecessary or incorrect.

elevatebart replied:
We check for an empty object here so when we get an empty object, is formatting of the output code. 
We return the code with a return in the middle instead of inline `{}`. Do you think my comparison with empty object will never match?

[EDIT] Objects are matched by reference in JS (duh!) my mistake. I need to fix this. good catch

elevatebart replied:
Done in 0dd3649ab1d8a773ac460eef2d8b2618b01f4ea0

elevatebart commented on cli/types/cypress.d.ts:
```diff
@@ -2717,15 +2717,17 @@ declare namespace Cypress {
      * @default {}
      */
-    component: Omit<ResolvedConfigOptions, TestingType>
+    component: Omit<ResolvedConfigOptions, TestingType> & { plugins?: TestingTypeFunctions }
```
Done in 38e76acf70d4a9e3432c5f79a9cb02e41eae8d19

elevatebart commented on packages/desktop-gui/src/settings/configuration.jsx:
```diff
@@ -175,12 +181,21 @@ const Configuration = observer(({ project }) => (
+          {project.hasE2EFunction ?
+            <>
+              <td><span className='function'>function</span></td>
+              <td>set in the <code>e2e</code> function if the {configFileFormatted(project.configFile)}</td>
```
Done in b1b2b8ecf608fcb2042ec0423cd174588705cd28

elevatebart commented on packages/server/lib/browsers/utils.ts:
```diff
@@ -132,6 +132,24 @@ async function executeBeforeBrowserLaunch (browser, launchOptions: typeof defaul
+    })
+  }
+
+  return pluginConfigResult
```
Done in 4c73fc1f80aa89bf5c8e5a93e79315fa4c8d5063

elevatebart commented on packages/server/lib/plugins/child/run_plugins.js:
```diff
@@ -140,8 +140,27 @@ const execute = (ipc, event, ids, args = []) => {
+    }
+
+    return finalFunction
+  }
```
Done in 5ec842e7eee1d7f950aa8323cd8cabf638860bef

elevatebart commented on packages/server/lib/plugins/util.js:
```diff
@@ -4,11 +4,62 @@ const debug = require('debug')('cypress:server:plugins')
+
+  if (typeof arg === 'object') {
+    if (_.isArray(arg)) {
+      return arg.map((argElement) => serializeArgument(argElement))
```
Done in 00c799d5f03467ce535043e459acd72c6d89ebab

elevatebart commented on packages/server/lib/saved_state.js:
```diff
@@ -49,13 +49,39 @@ const formStatePath = (projectRoot) => {
+          }
+
+          return projectRoot
+        })
```
Done in a8db11281f9235bcdd20cb491b7f99373b1fe908

cowboy replied:
Looks good! Please make sure it's tested, in case I made any mistakes in my code!

cowboy commented on cli/lib/cypress.js:
```diff
@@ -68,6 +68,19 @@ const cypressModuleApi = {
+   * type chekcing of the configuration given by users.
+   *
+   * @see ../types/cypress-npm-api.d.ts
+   * @param {Cypress.ConfigOptions} config
```
If we're trying to follow the pattern vite uses:

* https://github.com/vitejs/vite/blob/main/packages/vite/src/node/config.ts#L66-L68

Can we also export our `Cypress.ConfigOptions` type like they do:

* https://github.com/vitejs/vite/blob/main/packages/vite/src/node/config.ts#L58

So we can give our users options like they do?

* https://vitejs.dev/config/#config-intellisense

This way, people can use `defineConfig` if they want the intellisense, but they'll have the option to just use the type if they would like:

```ts
/**
 * @type {import('cypress').ConfigOptions}
 */
const config = {
  // config options
}

export default config
```

or

```ts
import type { ConfigOptions } from 'cypress'

const config: ConfigOptions = {
  // config options
}
export default config
```

elevatebart replied:
Yes, done

cowboy commented on cli/lib/cypress.js:
```diff
@@ -68,6 +68,19 @@ const cypressModuleApi = {
     },
   },
+
+  /**
+   * This function should not be used in pure JavaScript
+   * By the type of its argument it allows autocomplettion and
+   * type chekcing of the configuration given by users.
+   *
```
```suggestion
  /**
   * Type helper function to enable intellisense in compatible editors.
   *
```

cowboy commented on packages/server/lib/util/settings.js:
```diff
@@ -3,7 +3,36 @@ const Promise = require('bluebird')
+`
+}
+
+function tsCode (obj) {
```
If we export the `ConfigOptions` type, can we also update any example `cypress.config.ts` file we generate to feel a little more like standard Typescript, eg.

```ts
import type { ConfigOptions } from 'cypress'

const config: ConfigOptions = {
  
}

export default config
```

lmiller1990 commented on packages/server/lib/errors.js:
```diff
@@ -987,6 +1000,33 @@ const getMsgByType = function (type, arg1 = {}, arg2, arg3) {
+        https://on.cypress.io/setupNodeEvents
+        `
+
+    // TODO: update with vetted cypress language
```
Nice, this is great.

lmiller1990 commented on packages/server/lib/modes/run.js:
```diff
@@ -462,7 +462,7 @@ const iterateThroughSpecs = function (options = {}) {
 }
 
-const getProjectId = Promise.method((project, id) => {
+const getProjectId = Promise.method(async (project, id) => {
```
`async` doesn't look to be needed.

lmiller1990 commented on packages/server/lib/plugins/child/run_plugins.js:
```diff
@@ -171,13 +205,17 @@ const runPlugins = (ipc, pluginsFile, projectRoot) => {
-    if (plugins && typeof plugins.default === 'function') {
-      plugins = plugins.default
+    if (testingType === 'component') {
+      setupDevServerFunction = pluginsObject[testingType].setupDevServer
```
```suggestion
      setupDevServerFunction = pluginsObject.component.setupDevServer
```

lmiller1990 commented on packages/server/lib/plugins/util.js:
```diff
@@ -4,11 +4,64 @@ const debug = require('debug')('cypress:server:plugins')
   return _.pick(err, 'name', 'message', 'stack', 'code', 'annotated', 'type')
 }
 
+function serializeArgument (arg) {
```
This function might be a good candidate for some unit tests - not sure how thorough it is covered by the e2e tests.

lmiller1990 commented on packages/server/lib/plugins/util.js:
```diff
@@ -4,11 +4,64 @@ const debug = require('debug')('cypress:server:plugins')
+  return arg
+}
+
+function deserializeArgument (arg) {
```
Same here.

lmiller1990 commented on packages/server/lib/project-base.ts:
```diff
@@ -813,15 +828,15 @@ export class ProjectBase<TServer extends ServerE2E | ServerCt> extends EE {
+  async getProjectId (config?: Cfg): Promise<string> {
     await this.verifyExistence()
-    const readSettings = await settings.read(this.projectRoot, this.options)
+    const readSettings = config || await settings.read(this.projectRoot, this.options)
```
```suggestion
    const readSettings = this.getConfig() || await settings.read(this.projectRoot, this.options)
```

lmiller1990 commented on packages/server/lib/project-base.ts:
```diff
@@ -813,15 +828,15 @@ export class ProjectBase<TServer extends ServerE2E | ServerCt> extends EE {
   // These methods are not related to start server/sockets/runners
 
-  async getProjectId () {
+  async getProjectId (config?: Cfg): Promise<string> {
```
```suggestion
  async getProjectId (): Promise<string> {
```

lmiller1990 replied:
We should not need to pass around `config` anymore, just call `getConfig()` - this should be the source of truth.

lmiller1990 commented on packages/server/lib/saved_state.js:
```diff
@@ -49,17 +50,23 @@ const formStatePath = (projectRoot) => {
 
-    const cypressJsonPath = cwd('cypress.json')
+    return CYPRESS_CONFIG_FILES.reduce((acc, filename) => {
+      return acc.then((projectRoot) => {
```
I found this confusing - took me a while to figure out what was going on. Why not just use a for loop and synchronous operations? Less code, and it's a lot more simple (but still avoids mutation). Not tested, but something like this:

```js
    for (const filename of CYPRESS_CONFIG_FILES) {
      const cypressConfigPath = cwd(filename)
      const found = fs.pathExistsSync(cypressConfigPath)
      if (found) {
        const projectRoot = cwd()
        const fileName = 'state.json'
        return path.join(appData.toHashName(projectRoot), fileName)
      }
    }

    return path.join('__global__', fileName)
```

Could use reduce as well, but I think we want to return as soon as we find a config, so a for loop makes more sense.


Definitely easier to read without `async` and `then`.

lmiller1990 commented on packages/server/lib/util/require_async.ts:
```diff
@@ -0,0 +1,78 @@
+
+let requireProcess: cp.ChildProcess | null
+
+interface RequireAsyncOptions{
```
Nit: Linting - CI should be failing, if it's not, we should look into it. You should be able to configure VS Code to infer indentation based on the file's current indentation, though.

elevatebart replied:
Done in 87c1202ba409dcdae7ad3a0031177dc0eb7622cb

lmiller1990 commented on packages/server/lib/util/require_async.ts:
```diff
@@ -0,0 +1,78 @@
+  requireProcess = null
+}
+
+export default async function requireAsync (filePath: string, options: RequireAsyncOptions): Promise<any> {
```
```suggestion
export default async function requireAsync (filePath: string, options: RequireAsyncOptions): Promise<void> {
```

This function will never return a value.

elevatebart replied:
I think it will return a value. 
This is the purpose of the `resolve()` function.

It will return the exported object of the file.
Since we do not know the type of that value we say `any`.

I could also go for `T=any`. If we know the theoretical contents of a file this could be useful. Though in this case I would prefer the use of `as`

lmiller1990 commented on packages/server/lib/util/require_async_child.js:
```diff
@@ -0,0 +1,76 @@
+require('graceful-fs').gracefulify(require('fs'))
+const debug = require('debug')('cypress:server:require_async:child')
```
Nit: we really should be enforcing TS for any new files - this one should be easy to convert, but adding new files without types feel like a bit of a step backwards.

elevatebart replied:
It was initially a copy paste of a js file.
I will convert.

elevatebart replied:
Done in 12a20f9

elevatebart replied:
It is actually impossible since this file is called as a child process... reverting

lmiller1990 commented on packages/server/lib/util/settings.js:
```diff
@@ -93,12 +136,40 @@ module.exports = {
+      const evaluatedFiles = foundConfigFiles.filter((file) => file !== 'cypress.json')
+
+      // if we found more than one evaluated file (js or ts)
+      // we can't trust that migration has been doen properly
```
```suggestion
      // we can't trust that migration has been done properly
```

lmiller1990 commented on packages/server/test/cache_helper.js:
```diff
@@ -0,0 +1,19 @@
+const supportedConfigFiles = [
```
Should we use `CYPRESS_CONFIG_FILES` here?

Also pls make the new files TS - it will bring a lot of long term value, there's not reason not to.

elevatebart replied:
Done in bb00ece

lmiller1990 commented:
I found an edge case:

1. make config file

```ts
export default {
  "projectId": "ypt4pf",
  "baseUrl": "http://localhost:3500",
  viewportHeight: 700,
  "retries": {
    "runMode": 2,
    "openMode": 0
  }
}
```

2. open desktop gui
3. add `random.ts` with:

```ts
export const height = 999
```

4. update `cypress.config.ts`:

```ts
import { height } from "./random";

export default {
  "projectId": "ypt4pf",
  "baseUrl": "http://localhost:3500",
  viewportWidth: height,
  viewportHeight: 700,
  "retries": {
    "runMode": 2,
    "openMode": 0
  }
}
```

Desktop GUI won't refresh. Not sure if we can solve this or not, or if it's really a big deal.

elevatebart commented on packages/desktop-gui/src/projects/projects-api.js:
```diff
@@ -192,7 +193,21 @@ const openProject = (project) => {
+
+    config = _.omit(config, 'e2e', 'component')
+
+    updateConfig(config, hasE2Eproperty)
```
Done in d5909d7

elevatebart commented on packages/runner-ct/src/app/RunnerCt.tsx:
```diff
@@ -176,72 +177,91 @@ const RunnerCt = namedObserver('RunnerCt',
-            />
-          )}
+      <>
+        {renderWarnings()}
```
Done in 5c1cb6d

elevatebart commented on packages/runner-ct/src/app/WarningMessage.tsx:
```diff
@@ -0,0 +1,17 @@
+import styles from './WarningMessage.module.scss'
+
+interface WarningMessageProps{
+	warning: Warning
```
Done in 81f93b5

jennifer-shehane commented:
Converting to draft until ready for rereview to quiet notifications.

cowboy commented on packages/runner-ct/src/lib/state.ts:
```diff
@@ -55,6 +55,10 @@ const _defaults: Defaults = {
   callbackAfterUpdate: null,
 }
 
+export interface Warning{
```
Do we have a code formatter / linter built in to handle this kind of thing?

flotwig commented:
@lmiller1990 

> ## New module not detected
> 
>     1. make config file
> 
> 
> ```ts
> export default {
>   "projectId": "ypt4pf",
>   "baseUrl": "http://localhost:3500",
>   viewportHeight: 700,
>   "retries": {
>     "runMode": 2,
>     "openMode": 0
>   }
> }
> ```
> 
>     1. open desktop gui
> 
>     2. add `random.ts` with:
> 
> 
> ```ts
> export const height = 999
> ```
> 
>     1. update `cypress.config.ts`:
> 
> 
> ```ts
> import { height } from "./random";
> 
> export default {
>   "projectId": "ypt4pf",
>   "baseUrl": "http://localhost:3500",
>   viewportWidth: height,
>   viewportHeight: 700,
>   "retries": {
>     "runMode": 2,
>     "openMode": 0
>   }
> }
> ```
> 
> Desktop GUI won't refresh. Not sure if we can solve this or not, or if it's really a big deal.

This code is supposed to watch the config file and reload if it or any required files change, I believe? This seems like a bug: https://github.com/cypress-io/cypress/blob/69ebdbafeb21abcd19316f16256374130b51dff0/packages/server/lib/project-base.ts#L559-L561

lmiller1990 commented:
@flotwig the code you linked looks like it *should* be working. It's checking if `configFile` is not `false`, and watching it. I wonder if that was added after I raised the bug and is supposed to fix it. I'll have to give this another test, to see if it's been fixed or not, since @elevatebart is out for a while.

jennifer-shehane reviewed (changes requested):
Transfering the npm directories to the new cypress.config.js should be done in a separate PR considering this is not a breaking change. It's best to keep the PR to be the bare minimum required for the feature. The PR is already very large and it would help get a faster review since this requires every member of the team to review. Also any refactors like the markdown-renderer refactor should be in a separate PR. 

I've left comments within the code for review. Please expand all hidden comments, there is a lot. 

Outside of those, some observations while manually testing:

1. When the error is shown about duplicate config, the loading spinner for the projects area is still shown and spinning. 

	
	<img width="961" alt="Screen Shot 2021-08-24 at 9 17 17 AM" src="https://user-images.githubusercontent.com/1271364/130634038-cf3999d5-bd4a-4601-bf3f-a92fd046a142.png">

2. Can we reword this message instead of saying 'retry again' to explicitly tell them what to do. 

	> Please remove one of the two, quit and relaunch Cypress. 

3. So I just pasted the JSON from the `cypress.json` file directly into the `cypress.config.js` file, because I kind of expect people to just do this without reading any of the migration guide and I get this error. I'm not sure if the error here can be better. Is there a way to detect that they just have JSON in the file and explain to them that if it's pure JSON that they need to migrate to export, etc - link to migration guide?

	<img width="961" alt="Screen Shot 2021-08-24 at 9 26 34 AM" src="https://user-images.githubusercontent.com/1271364/130634885-f05ed0ee-9e88-4454-bc26-3bf3112f9be2.png">

4. When I added the export as shown in the PR description, I got this error about unexpected token, 'export'. 
	
	<img width="436" alt="Screen Shot 2021-08-24 at 10 32 44 AM" src="https://user-images.githubusercontent.com/1271364/130645949-f840ada7-5116-4361-8b61-12274bf34dcd.png">

	
	<img width="846" alt="Screen Shot 2021-08-24 at 10 27 04 AM" src="https://user-images.githubusercontent.com/1271364/130645078-24364762-9dfd-41d0-b551-98599b014366.png">

5. Then I tried to switch to the `defineConfig` method I'd seen in the config.ts in this repo and it throws an error again. I guess my bit of takeaway from this is that these error messages are not helpful at all. They should say something like:

	> There's an error in your config file at the following location: ...
	> Make sure that your config file is structured like this:
	> ...minimum required code structure for file....
	> link to guide where it explains how to write in config file. 

And secondly apparently I have no idea how to write this config file because this PR description is not up to date. Please keep the PR description up to date so people can find the proper documentation to fully review without having to search through outside resources.
	
	<img width="515" alt="Screen Shot 2021-08-24 at 10 29 33 AM" src="https://user-images.githubusercontent.com/1271364/130645482-10513f7e-f090-4778-bfad-cc303d24bda8.png">

	<img width="798" alt="Screen Shot 2021-08-24 at 10 30 17 AM" src="https://user-images.githubusercontent.com/1271364/130645572-b17a8f2a-3a60-4c82-b0a6-c35815404194.png">

6. Was just thinking through another case where the cypress.json and `cypress.config.js` will co-exist together. If someone runs Cypress on 8.x (where `cypress.config.js` is scaffolded), then quickly runs Cypress on 8.3.0 to test some change in the update, it will scaffold the `cypress.json` back again. May be annoying for a few people trying to track down some regression for it to scaffold a diff file for each back and forth.

elevatebart commented:
Thank you @jennifer-shehane 
@lmiller1990 is right, great review once again.

Two details I wanted whomever is taking over the pull request need to know:
- I did the migration of npm packages in this PR to use them as end to end tests. The PR is already humongous, so tried to make it complete and coherent. If we add these tests in a separate PR, it will be more work for all of us.
- The failure on the `export default` is actually [an issue in ts-node](https://github.com/TypeStrong/ts-node/issues/1007). When TypeScript is installed, one cannot have a `cypress.config.js` file using `export default`. This file has to use `module.exports = `. The team at ts-node are still working on it but want to avoid breaking changes as much as possible. Should we make it a clearer failure? Should we wait for their fix?

chrisbreiding commented on packages/server/test/unit/plugins/util_spec.js:
```diff
@@ -55,6 +55,115 @@ describe('lib/plugins/util', () => {
+    })
+
+    context('#arguments-deserialization', () => {
+      it('#on should deserialise function strings into function', function () {
```
Same as with `#send`, `#on` should be moved out into a `describe`.

chrisbreiding reviewed (changes requested):
Jennifer and Lachlan did pretty thorough reviews, so I don't have too much to add besides some nitpicks.
